test/zdtm/static/pidfd_dead.c

```diff
@@ -0,0 +1,157 @@
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/utsname.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc = "Check C/R of pidfds that point to dead processes\n";
+const char *test_author = "Bhavik Sachdev <b.sachdev1904@gmail.com>";
+
+static int pidfd_open(pid_t pid, unsigned int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int pidfd_send_signal(int pidfd, int sig, siginfo_t* info, unsigned int flags)
+{
+	return syscall(__NR_pidfd_send_signal, pidfd, sig, info, flags);
+}
+
+static int open_pidfd_pair(int pidfd[2], int pid)
+{
+	pidfd[0] = pidfd_open(pid, 0);
+	if (pidfd[0] < 0)
+		return 1;
+	pidfd[1] = pidfd_open(pid, 0);
+	if (pidfd[1] < 0)
+		return 1;
+	return 0;
+}
+static int compare_pidfds(int pidfd[2])
+{
+	/*
+	* After linux 6.9 we can compare inode numbers
+	* to determine if two pidfds point to the same process.
+	* While the inode number may change before and after C/R
+	* pidfds pointing to the same pid should have the same inode number.
+	*/
+	struct statx stats[2];
+	statx(pidfd[0], "", AT_EMPTY_PATH, STATX_ALL, &stats[0]);
+	statx(pidfd[1], "", AT_EMPTY_PATH, STATX_ALL, &stats[1]);
+	if (stats[0].stx_ino != stats[1].stx_ino)
+		return 1;
+	return 0;
+}
+
+static struct utsname buf;
+
+int main(int argc, char* argv[])
+{
+	#define READ 0
+	#define WRITE 1
+
+	int child, ret, gchild, p[2];
+	int cpidfd[2], gpidfd[2];
+	unsigned major, minor;
+
+	test_init(argc, argv);
+
+	ret = uname(&buf);
+	if (ret)
+		return 1;
+
+	if (sscanf(buf.release, "%u.%u", &major, &minor) != 2)
+		return 1;
+
+	if ((major <= 6) && (minor < 9)) {
```

Comment:
Instead of doing a version-based check, you can detect pidfdfs feature presence using `fstat` syscall and checking file system magic.

Something like that:
```
static long get_fs_type(int lfd)
{
	struct statfs fst;

	if (fstatfs(lfd, &fst)) {
		pr_perror("Unable to statfs fd %d", lfd);
		return -1;
	}
	return fst.f_type;
}

get_fs_type(pidfd) == PID_FS_MAGIC

```

Reply:
Done!